dhcpcd: enable sandboxing options

Izorkin opened the pull request:
###### Description of changes
Enable sandboxing options.
Result:
```
systemd-analyze security dhcpcd
...
→ Overall exposure level for dhcpcd.service: 2.9 OK 🙂
```
cc @SuperSandro2000

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Izorkin committed:
dhcpcd: enable sandboxing options

SuperSandro2000 commented on nixos/doc/manual/release-notes/rl-2305.section.md:
```diff
@@ -94,6 +94,8 @@ In addition to numerous new and upgraded packages, this release has the followin
 
   And backup your data.
 
+- `services.dhcpcd` is now started with additional systemd sandbox/hardening options for better security.
```
Is this relevant? Do we expect some configurations to break? If not this can be removed.

Izorkin replied:
No, a simple notification. Deleted.

SuperSandro2000 commented on nixos/modules/services/networking/dhcpcd.nix:
```diff
@@ -226,6 +226,38 @@ in
+            RestrictSUIDSGID = true;
+            RemoveIPC = true;
+            PrivateMounts = true;
+            # System Call Filtering
```
```suggestion
```

SuperSandro2000 commented on nixos/modules/services/networking/dhcpcd.nix:
```diff
@@ -226,6 +226,38 @@ in
             ExecStart = "@${dhcpcd}/sbin/dhcpcd dhcpcd --quiet ${optionalString cfg.persistent "--persistent"} --config ${dhcpcdConf}";
             ExecReload = "${dhcpcd}/sbin/dhcpcd --rebind";
             Restart = "always";
+            # Proc filesystem
+            ProcSubset = "all";
+            ProtectProc = "invisible";
+            # Access write directories
+            UMask = "0027";
+            # Capabilities
+            CapabilityBoundingSet = [ "CAP_NET_ADMIN" "CAP_NET_BIND_SERVICE" "CAP_NET_RAW" "CAP_SETGID" "CAP_SETUID" "CAP_SYS_CHROOT" ];
+            # Security
+            NoNewPrivileges = true;
+            # Sandboxing
```
```suggestion
            ProcSubset = "all";
            ProtectProc = "invisible";
            UMask = "0027";
            CapabilityBoundingSet = [ "CAP_NET_ADMIN" "CAP_NET_BIND_SERVICE" "CAP_NET_RAW" "CAP_SETGID" "CAP_SETUID" "CAP_SYS_CHROOT" ];
            NoNewPrivileges = true;
```

Izorkin replied:
I usually leave comments like this everywhere to make it easier to navigate the parameters. For example in `nginx` and `mastodon` services.

blitz commented on nixos/modules/services/networking/dhcpcd.nix:
```diff
@@ -226,6 +226,38 @@ in
+            PrivateUsers = false;
+            ProtectHostname = true;
+            ProtectClock = true;
+            ProtectKernelTunables = false;
```
nit: For the ones you opt out of, a short comment why would be interesting.

Izorkin replied:
When `ProtectKernelTunables` is active, there is an error in the `dhcpcd` log:
```
dhcpcd[14944]: ipv6nd_applyra: Read-only file system
```

Izorkin replied:
For example, these changes work on my test virtual machine:
```diff
diff --git a/nixos/modules/services/networking/dhcpcd.nix b/nixos/modules/services/networking/dhcpcd.nix
index 4cf86653205..bcd6361ec1b 100644
--- a/nixos/modules/services/networking/dhcpcd.nix
+++ b/nixos/modules/services/networking/dhcpcd.nix
@@ -230,13 +230,14 @@ in
             ProcSubset = "all";
             ProtectProc = "invisible";
             # Access write directories
+            ReadWritePaths = [ "/etc/resolv.conf" "/var/db/dhcpcd" ];
             UMask = "0027";
             # Capabilities
             CapabilityBoundingSet = [ "CAP_NET_ADMIN" "CAP_NET_BIND_SERVICE" "CAP_NET_RAW" "CAP_SETGID" "CAP_SETUID" "CAP_SYS_CHROOT" ];
             # Security
             NoNewPrivileges = true;
             # Sandboxing
-            ProtectSystem = true;
+            ProtectSystem = "strict";
             ProtectHome = true;
             PrivateTmp = true;
             PrivateDevices = true;
```
But these parameters do not work in the `nixosTests.networking.scripted.dhcpDefault` test. I can't figure out why. That's why I deleted these changes.

fpletz reviewed (dismissed):
This might break the setup of people using `networking.dhcpcd.runHook`. I don't think we should go ahead with this without at least an entry in the release notes and an option to disable hardening.

Izorkin commented:
> This might break the setup of people using `networking.dhcpcd.runHook`. I don't think we should go ahead with this without at least an entry in the release notes and an option to disable hardening.

Updated PR.
Now when using `networking.dhcpcd.runHook` these settings are not applied.
Added relese-notes.

Izorkin commented:
Resolving conflicts.

Izorkin commented:
Rebased PR.

Izorkin commented:
Resolved conflicts.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/1993

Izorkin commented:
On my virtual machines, PR does not cause obvious problems, it works fine :)

Izorkin commented:
Rebased PR.

Izorkin commented:
It is possible to include this PR in the 23.05 release?

Izorkin commented:
Small update and rebase PR.

SuperSandro2000 reviewed (approved):
I didn't test this on my machine but if the author tested it fine for me

SuperSandro2000 commented:
I am no longer using dhcpcd

Izorkin commented:
Rebased PR.
What is needed to get this PR merged?

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/2020

flokli commented:
From https://github.com/NixOS/nixpkgs/pull/208780#pullrequestreview-1238673041, slightly highlighted:

> This might break the setup of people using `networking.dhcpcd.runHook`. I don't think we should go ahead with this without at least an entry in the release notes <ins>and an option to disable hardening</ins>.

This is still missing said option, as well as a mention of it in the release notes.

Izorkin commented:
> This is still missing said option, as well as a mention of it in the release notes.

Doesn't this line disable sandbox mode?
```
          } // lib.optionalAttrs (cfg.runHook == "") {
```

flokli commented:
Ah, now I understand, you don't apply sandboxing at all.

flokli commented:
I still hope we can get rid of scripted networking altogether, but today is not that day, so sure, let's add the sandboxing. Thanks for the PR!

Izorkin commented:
Thanks!

vcunat commented:
FYI, this clashes with https://github.com/NixOS/nixpkgs/pull/336988
(right now in merge from master to staging-next)

Izorkin commented:
Sorry, I didn't know about the other PR :(
